### PR TITLE
Add many Readers

### LINKED ISSUE #2509: doctest memory increasing
**Describe what maintenance you would like added.**

The current usage of doctest has increasing memory.  #2496 has increased the number of doctests to the point that the CI runs out of memory.  Locally, the memory increases >10 GB on some commit refs.  Currently, none of the Plotters are being closed, and thus memory increases each time a plot is made.  Confirmation of this is below.  However, normal doctest does not allow for convenient use of teardown code.  And we'd have to add `pyvista.all_close()` to every doctest.

`pytest-doctestplus` can fix this issue while providing more flexible doctesting for our uses.  It allows using pytest fixtures, which enables for clean teardown code to be run after each doctest without adding fluff to each doctest.

When using `pytest-doctestplus` with the fixture below eliminates the memory buildup seen, which confirms that the plotters aren't closing properly today.

```py
@pytest.fixture(autouse=True)
def autoclose_plotters():
    yield
    pyvista.close_all()
```

### Downsides

- Potentially breaks usage with base doctest.  The option flag `NUMBER` is not supported, but `FLOAT_CMP` can be used.  `FLOAT_CMP` rounds to nearest whereas `NUMBER` rounds down.  Switching would prevent using normal doctest. This could potentially be avoided by using ELLIPSES instead, but it might have unintended edge cases.
- I'm not sure if this impacts `check_doctest_names`

### PR DESCRIPTION
**Pull request by MatthewFlamm**
### Overview

This PR will attempt to add all the readers used in the standard_reader_routine in the fileio module to enable #2494.

### Details

A followup PR will be made to align the read vs. Reader usage.

A checklist of readers from #2494 that I will track here:

- [x] vtkBMPReader
- [x] vtkDEMReader
- [x] vtkJPEGReader
- [x] vtkMetaImageReader
- [x] vtkNrrdReader
- [x] vtkPNGReader
- [x] vtkPNMReader
- [x] vtkSLCReader
- [x] vtkTIFFReader
- [x] vtkGLTFReader  <- no data
- [x] vtkHDRReader
- [x] vtkFluentReader  <- no data
- [x] vtkMFIXReader  <- no data
- [x] vtkPTSReader  <- no data
- [x] vtkAVSucdReader
- [x] vtkSegYReader  <- no data
- [x] vtkHDFReader

TODO:
- [x] tests
- [x] add all to pyvista namespace
- [x] add to documentation
- [x] ~~add instantiation tests for readers without available data~~

### Bonus

Resolves #2509

### CODE REVIEW

**Comment by MatthewFlamm** [comment]
I have implemented all the Readers that I can find datasets for in vtk-data.  I won't be able to test the readers that have no suitable data.  I will indicate which ones in the description above.

I will add those Readers and then start adding in true tests.

**Comment by MatthewFlamm** [comment]
I'm assuming I added too many plots for the docstring test.  Should I be skipping testing all doctests that download and plot datasets?

**Comment by adeak** [comment]
When I run the doctests locally I see a pretty steep rise in memory need. That has to be something we should fix, right? There's no reason why anything should persist across doctest runs, so there should be no accumulation of memory.

I first thought that this is related to [doctests sharing the module global namespace](https://github.com/pytest-dev/pytest/discussions/8791), but [at least stdlib doctest says](https://docs.python.org/3/library/doctest.html#what-s-the-execution-context)

> By default, each time [doctest](https://docs.python.org/3/library/doctest.html#module-doctest) finds a docstring to test, it uses a shallow copy of M’s globals, so that running tests doesn’t change the module’s real globals, and so that one test in M can’t leave behind crumbs that accidentally allow another test to work. This means examples can freely use any names defined at top-level in M, and names defined earlier in the docstring being run. Examples cannot see names defined in other docstrings.

So I wonder why allocated memory keeps building up...

**Comment by MatthewFlamm** [comment]
Other than figuring out the doctest issue, this is ready to go.

I decided not to implement tests for readers without available data.  One reader hung when using fake data even when just setting the file name.  So please pay particular attention to those 5 implementations during review.

**Comment by akaszynski** [comment]
> When I run the doctests locally I see a pretty steep rise in memory need (2 GB at 36% execution, and counting). That has to be something we should fix, right? There's no reason why anything should persist across individual doctest executions, so there should be no accumulation of memory.

Locally, I see 8GB + of resident memory usage, which exceeds the GitHub runner [Supported runners and hardware resources](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources). Even adding a `gc.collect` into a `conftest.py` doesn't help; the references are sticking around between tests.

**Comment by akaszynski** [comment]
Played around with this for a bit. I think given the limitations of `doctest` and `pytest`, we're going to have to provide ourselves with a workaround. Might be one of the following:
- Run `pytest --doctest-modules pyvista` on individual modules within pyvista. This will work, but will require some maintenance if modules are added or removed. For example:
  ```
  python -m pytest --doctest-modules pyvista/plotting
  python -m pytest --doctest-modules pyvista/utilities
  ...
  ```
- Use @adkea's [check_doctest_names.py](https://github.com/pyvista/pyvista/blob/main/tests/check_doctest_names.py), but we'll miss out on checking output.

**Comment by MatthewFlamm** [comment]
I can confirm locally that using `pytest-doctestplus` with a teardown fixture that uses `pyvista.close_all` fixes memory buildup during doctests.

We have to make a decision before converting to `pytest-doctestplus`, because it breaks compatibility with the normal doctest.  One example is that the normal doctest won't use fixtures, but this is okay as long as the fixtures we write aren't required to run the tests.  

More important is that `pytest-doctestplus` changes some of the behaviors when comparing values such as floats or ELLIPSIS.  For example, float values will now be rounded to nearest instead of rounded down.  Modifying our doctests to accommodate will break the tests on the normal doctest.

**Comment by MatthewFlamm** [comment]
I've started #2509 so we can discuss this there.

**Comment by MatthewFlamm** [comment]
If 5776ca8677ae76a25214222ad4dd090698a231e5 fixes the doctest memory issue, we can revert the `# doctest: +SKIP` commits.

**Comment by adeak** [comment]
```
pyvista/utilities/reader.py::pyvista.utilities.reader.HDRReader PASSED          [ 96%]
pyvista/utilities/reader.py::pyvista.utilities.reader.TIFFReader PASSED          [ 98%]
```
:+1:

**Review by akaszynski — APPROVED** [review]
LGTM. I made one final change to ignore coverage of `conftest.py`. If that goes through, feel free to merge.

**Comment by adeak** [comment]
@MatthewFlamm earlier you said

> I decided not to implement tests for readers without available data. One reader hung when using fake data even when just setting the file name. So please pay particular attention to those 5 implementations during review.

Is this still relevant? Which are the 5 readers that did that?

File: .coveragerc [hunk 1]
```diff
@@ -3,3 +3,4 @@ omit =
     pyvista/ext/coverage.py
     # kept for backwards compatibility:
     pyvista/plotting/theme.py
+    pyvista/conftest.py
```
Comment (adeak) [inline .coveragerc:6]:
```suggestion
    pyvista/ext/coverage.py
    pyvista/conftest.py
    # kept for backwards compatibility:
    pyvista/plotting/theme.py
```

File: pyvista/conftest.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
+"""Close all plotters to help control memory usage for our doctest."""
```
Comment (adeak) [inline pyvista/conftest.py:1]:
```suggestion
"""Close all plotters to help control memory usage for our doctests."""
```

File: pyvista/examples/downloads.py [hunk 3]
```diff
@@ -3612,22 +3636,24 @@ def download_lucy(load=True):  # pragma: no cover
     return _download_and_read('lucy.ply', load=load)
 
 
-def download_can(partial=False):  # pragma: no cover
+def download_can(partial=False, load=True):  # pragma: no cover
     """Download the can dataset mesh.
 
-    Original downloaded from Testing/Data/FileSeriesat from paraview.org. Used
+    Original downloaded from Testing/Data/FileSeries at from paraview.org. Used
```
Comment (adeak) [inline pyvista/examples/downloads.py:3642]: 1. What does "at from paraview.org" mean? Typo of "at paraview.org"?
2. What does "Testing/Data/FileSeries at paraview.org" mean? I can't find that.

File: pyvista/examples/downloads.py [hunk 4]
```diff
@@ -3612,22 +3636,24 @@ def download_lucy(load=True):  # pragma: no cover
...
     for testing hdf files.
 
     Parameters
     ----------
     partial : bool, optional
-        Load part of the dataset. Defaults to to ``False`` and
-        filename will be returned.
+        Load part of the dataset. Defaults to ``False``.
+    load : bool, optional
```
Comment (adeak) [inline pyvista/examples/downloads.py:3649]:
```suggestion

    load : bool, optional
```

File: pyvista/examples/downloads.py [hunk 5]
```diff
@@ -3780,3 +3808,51 @@ def download_dicom_stack(load: bool = True) -> Union[pyvista.UniformGrid, str]:
...
+    >>> from pyvista import examples
+    >>> dataset = examples.download_parched_canal_4k()
+    >>> dataset.plot(cpos="xy")
+
+    """
+    return _download_and_read("parched_canal_4k.hdr", texture=True, load=load)
+
+
+def download_cells_nd(load=True):  # pragma: no cover
+    """Download dataset.
```
Comment (adeak) [inline pyvista/examples/downloads.py:3838]: Can we qualify this dataset somehow?

File: pyvista/utilities/reader.py [hunk 6]
```diff
@@ -1565,23 +1611,304 @@ class DICOMReader(BaseReader):
...
+
+
+class MFIXReader(BaseReader):
+    """MFIXReader for .res files."""
+
+    _class_reader = _vtk.vtkMFIXReader
+
+
+class SegYReader(BaseReader):
+    """SegYReader for .sgy and segy files."""
```
Comment (adeak) [inline pyvista/utilities/reader.py:1867]:
```suggestion
    """SegYReader for .sgy and .segy files."""
```

File: pyvista/examples/downloads.py (hunk 3 above)
Comment (MatthewFlamm) [reply pyvista/examples/downloads.py:3642]: I don't know.  I thought this was a typo, but it doesn't make sense in the changed form either.  kitware moved a lot of their testing data to a new site https://data.kitware.com/. So, this bread crumb might be stale.  Maybe we should just say "Original downloaded from Paraview."

File: pyvista/examples/downloads.py (hunk 5 above)
Comment (MatthewFlamm) [reply pyvista/examples/downloads.py:3839]: I have have no idea what it is, just that the data was available in vtk-data.  I can plot it and make a description of what it looks like.

File: pyvista/examples/downloads.py (hunk 3 above)
Comment (adeak) [reply pyvista/examples/downloads.py:3642]: Yeah, I can't seem to find it on the new site either. I agree we should use something generic then. Should it be Paraview or Kitware then?

Comment (akaszynski) [reply pyvista/examples/downloads.py:3642]: This is my fault. I should have provided a direct link.

Just cite Kitware

```suggestion
    File obtained from `Kitware <https://www.kitware.com/>`_. Used
```

**Comment by MatthewFlamm** [comment]
I will wait for the CI to run and then merge tomorrow if I have time if there are no more comments (or planned reviews)

**Pull request merged**